### ISSUE TITLE
DINOv2 performance vs DINO

### ISSUE THREAD

**Issue by rbareja25** [facebookresearch/dinov2#272]
hi,

I am working on a custom medical imaging binary classification task. I am comparing performances between DINO(VIT base) and DINOv2(VIT large)  after training DINO and DINOv2 models, with eval_linear code. With DINO using VIT base I am getting 85% classification accuracy on the test dataset, with the model trained for 100 epochs, whereas with DINOv2 trained for 300 epochs from training_362499/teacher_checkpoint model, I am getting only 50% accuracy on the same test dataset.

I know it's hard to know the reason for this but any ideas that I could try or something that I might be doing wrong?
Any suggestions would be appreciated.

Thanks,
Rohan

**Comment by vladchimescu** [comment]
Hi @rbareja25,
I am experiencing similar issues, i.e. I am struggling to reproduce DINO results using DINOv2. As you point out, this may be driven by hyper parameters.

In our case we used exactly the same hyper parameters as in DINO. We also conducted an ablation study, switching off KoLeo and iBOT losses without any success. 

@qasfb @patricklabatut Would you expect to get comparable results to DINO if you trained DINOv2 with ibot_loss_weight = 0, koleo_loss_weight = 0 and identical hyperparameters?

**Comment by alexaatm** [comment]
Hi @vladchimescu @rbareja25 I faced the same issue with segmentation based on dino vs dinov2 features. Then another look up in dinov2 paper reveals the authors are aware of such a discrepency:

from dinov2 paper: [dinov2](https://arxiv.org/pdf/2309.16588.pdf)

> When used to extract features, it delivers disappointing performance, only on par with supervised alternative backbones in this scenario. This suggests that DINOv2 behaves differently than DINO. The investigation described in this work notably exposes the presence of artefacts in the feature maps of DINOv2 that were not present in the first version of this model


[…]

`We note that we have not been able to fully determine which aspects of the training led to the appearance of artifacts in DINOv2 but not in DINO, but Fig. 4 suggests that scaling the model size beyond ViT-L, and longer training length may be possible causes`


Hope it helps to clear up some confusion.

P.S. i haven't tried the dinov2 with registers though, if anything changes, I'll report here:)

**Comment by vladchimescu** [comment]
Hi @alexaatm,
thank you for pointing this out. 

@qasfb @patricklabatut I believe this is an implementation issue, however. We set KoLeo and iBOT loss weights to zero and we could not reproduce the good performance of DINOv1, despite using the same hyperparameters. 

We haven't tried ViTs with registers as we're training DINO + ViT-S. The paper that you cite @alexaatm suggests that registers are needed for large ViTs trained on massive datasets.

**Comment by qasfb** [comment]
Hey ! Can you say a bit more about your setup, how many GPUs, what batch size and what config in general ?

**Comment by vladchimescu** [comment]
@qasfb Sure! We are using NVIDIA A100 GPUs (80 GB GPU memory). For the ViT-S backbone, we only need 1 GPU, but we also tried FSDP with 2 GPUs.

We used the same batch size (256) as for DINO (Caron et al, 2021). We found that for our custom dataset, having batch size <= 256 was beneficial. 

To reproduce our DINO baseline, we switched off iBOT and KoLeo losses and used the same student and teacher hyperparameters that we had previously used for DINO. I'm pasting the config file below:
```
MODEL:
  WEIGHTS: ''
compute_precision:
  grad_scaler: true
  teacher:
    backbone:
      sharding_strategy: SHARD_GRAD_OP
      mixed_precision:
        param_dtype: fp16
        reduce_dtype: fp16
        buffer_dtype: fp32
    dino_head:
      sharding_strategy: SHARD_GRAD_OP
      mixed_precision:
        param_dtype: fp16
        reduce_dtype: fp16
        buffer_dtype: fp32
    ibot_head:
      sharding_strategy: SHARD_GRAD_OP
      mixed_precision:
        param_dtype: fp16
        reduce_dtype: fp16
        buffer_dtype: fp32
  student:
    backbone:
      sharding_strategy: SHARD_GRAD_OP
      mixed_precision:
        param_dtype: fp16
        reduce_dtype: fp16
        buffer_dtype: fp32
    dino_head:
      sharding_strategy: SHARD_GRAD_OP
      mixed_precision:
        param_dtype: fp16
        reduce_dtype: fp32
        buffer_dtype: fp32
    ibot_head:
      sharding_strategy: SHARD_GRAD_OP
      mixed_precision:
        param_dtype: fp16
        reduce_dtype: fp32
        buffer_dtype: fp32
dino:
  loss_weight: 1.0
  head_n_prototypes: 10000
  head_bottleneck_dim: 256
  head_nlayers: 3
  head_hidden_dim: 2048
  koleo_loss_weight: 0
ibot:
  loss_weight: 0
  mask_sample_probability: 0.5
  mask_ratio_min_max:
  - 0.1
  - 0.5
  separate_head: false
  head_n_prototypes: 10000
  head_bottleneck_dim: 256
  head_nlayers: 3
  head_hidden_dim: 2048
train:
  batch_size_per_gpu: 256
  output_dir: .
  saveckp_freq: 10
  seed: 0
  num_workers: 12
  OFFICIAL_EPOCH_LENGTH: 2204
  cache_dataset: true
  centering: "centering" # or "sinkhorn_knopp"
student:
  arch: vit_small
  patch_size: 16
  drop_path_rate: 0
  layerscale: 1.0e-05
  drop_path_uniform: true
  pretrained_weights: ''
  ffn_layer: "mlp"
  block_chunks: 0
  qkv_bias: true
  proj_bias: true
  ffn_bias: true
teacher:
  momentum_teacher: 0.9995
  final_momentum_teacher: 1
  warmup_teacher_temp: 0.01
  teacher_temp: 0.04 # TODO this should be set to 0.04 (!)
  warmup_teacher_temp_epochs: 30
optim:
  epochs: 200
  weight_decay: 0.04
  weight_decay_end: 0.4
  base_lr: 0.001  # learning rate for a batch size of 256
  lr: 0.  # will be set after applying scaling rule
  warmup_epochs: 20
  min_lr: 1.0e-06
  clip_grad: 3.0
  freeze_last_layer_epochs: 3
  scaling_rule: multiple_of_256
  patch_embed_lr_mult: 0.2
  layerwise_decay: 0.9
  adamw_beta1: 0.9
  adamw_beta2: 0.999
crops:
  global_crops_scale:
  - 0.32
  - 1.0
  local_crops_number: 8
  local_crops_scale:
  - 0.05
  - 0.32
  global_crops_size: 224
  local_crops_size: 96
evaluation:
  eval_period_iterations: 12500
```

**Comment by qasfb** [comment]
How did you pick the hyperparameters in this config ?
I see layerwise decay 0.9 and momentum teacher 0.9995, and i'm pretty sure these were not in DINO.

**Comment by vladchimescu** [comment]
@qasfb The momentum teacher is from the original DINO. Indeed, layer scale was not part of the original DINO and we also ran an experiment without layer scale, which didn't make a difference. 

I am not sure what `layerwise_decay` was and I left it as in one of the config files that I found in this repo. How can I switch off `layerwise_decay`? Set it to 0?

Do you have any other pointers regarding what is different compared to vanilla DINO?

**Comment by qasfb** [comment]
To disable: layerwise decay = 1.0 
Otherwise I don't know, I have not tried to reproduce DINO with this codebase so I don't know for sure if it would work well.

**Comment by ironb25** [comment]
Hi @vladchimescu - Were you able to improve DINOV2 performance?Could you share suggestions?

**Comment by ironb25** [comment]
> How did you pick the hyperparameters in this config ? I see layerwise decay 0.9 and momentum teacher 0.9995, and i'm pretty sure these were not in DINO. Similarly the layerscale does not happen in DINO either.

I see momentum teacher is present in DINO.
<img width="762" alt="Screen Shot 2024-02-15 at 11 00 13 AM" src="https://github.com/facebookresearch/dinov2/assets/65127651/84f95554-3900-4e7a-995e-51105e300574">

**Comment by wangh09** [comment]
Hey guys, I'm experiencing the same issue when trying to reproduce dinov1 with dinov2 codebase. And I happened to find out that the async centering calculation may be the cause. I can reproduce DINOv1 training when switched to sync-op. While async-op seems to work well when both dino and ibot losses are on.

**Comment by ayushnangia** [comment]
any clue about what happend?

**Comment by rbareja25** [comment]
No, I have not been able to reproduce results even after doing registers, gradient accumulation, etc .Maybe fine tuning the ImageNet weights would be a better strategy.

**Comment by arbelhizmi** [comment]
Hi, I'm also trying to train Dinov2 on medical dataset (crop images from a tissue around cell centers). No matter what configuration I try - ViT-B, ViT-S, lower-higher learning rates, with/without Koleo & iBOT it seems like the model is getting into collapse mode (where the student output the same output for any input (so the loss is minimized, but the model doesn't really learns anything useful) - the loss drop from start value of 8-10 to 0.05 after a single epoch (a full pass on the entire dataset). No matter what I try, I always getting into this collapse mode.
I'm attaching example of one of my loss curves

![Image](https://github.com/user-attachments/assets/00828f0d-bab8-413b-962d-e4eefa9390c4)

@qasfb - the value of 0.9 fir layerwise_decay is from the official repository of Dinov2 under `dinov2/configs/ssl_default_config.yaml`.

I'm training with 8 GPUs and I'm using Fabric module of lightning (in 'ddp' mode, not fully-sharded-data-parallel, but it's not suppose to matter anyway)

If anyone has an idea of how to solve this - I'll be very happy to hear!
